File: exp/exact-replay/python/macho_normalizer.py
import struct
import sys
from dataclasses import dataclass
from pathlib import Path


LC_SEGMENT = 0x1
LC_SEGMENT_64 = 0x19
LC_CODE_SIGNATURE = 0x1D


class MachOFormatError(ValueError):
    pass


@dataclass(frozen=True)
class CodeSignature:
    command_offset: int
    dataoff: int
    datasize: int


@dataclass(frozen=True)
class LinkeditSegment:
    command_offset: int
    segname: str
    vmsize_offset: int
    filesize_offset: int
    fileoff: int
    filesize: int
    vmsize: int


@dataclass(frozen=True)
class MachOImage:
    bits: int
    endian: str
    header_size: int
    ncmds: int
    sizeofcmds: int
    commands_end: int
    file_length: int
    signature: CodeSignature
    linkedit: LinkeditSegment


def _reject(message: str) -> None:
    raise MachOFormatError(f"invalid thin Mach-O: {message}")


def _format(data: bytes) -> tuple[int, str, str, int]:
    if len(data) < 4:
        _reject("truncated magic")
    magic = data[:4]
    formats = {
        bytes.fromhex("cffaedfe"): (64, "little", "<", 32),
        bytes.fromhex("feedfacf"): (64, "big", ">", 32),
        bytes.fromhex("cefaedfe"): (32, "little", "<", 28),
        bytes.fromhex("feedface"): (32, "big", ">", 28),
    }
    if magic not in formats:
        _reject(f"unsupported magic {magic.hex()}")
    return formats[magic]


def _unpack(data: bytes, order: str, layout: str, offset: int) -> tuple[int, ...]:
    parser = struct.Struct(f"{order}{layout}")
    if offset < 0 or offset + parser.size > len(data):
        _reject(f"truncated {layout} at {offset}")
    return parser.unpack_from(data, offset)


def _segment_name(data: bytes, offset: int) -> str:
    return data[offset : offset + 16].split(b"\0", 1)[0].decode("ascii", errors="strict")


def parse_macho(data: bytes) -> MachOImage:
    bits, endian, order, header_size = _format(data)
    if len(data) < header_size:
        _reject("truncated header")
    ncmds, sizeofcmds = _unpack(data, order, "II", 16)
    commands_end = header_size + sizeofcmds
    if commands_end > len(data):
        _reject("load command region exceeds the file")

    signatures: list[CodeSignature] = []
    linkedits: list[LinkeditSegment] = []
    command_offset = header_size
    expected_segment = LC_SEGMENT_64 if bits == 64 else LC_SEGMENT
    minimum_segment_size = 72 if bits == 64 else 56
    for index in range(ncmds):
        if command_offset + 8 > commands_end:
            _reject(f"truncated load command {index}")
        command, command_size = _unpack(data, order, "II", command_offset)
        if command_size < 8 or command_size % 4:
            _reject(f"invalid cmdsize for load command {index}")
        command_end = command_offset + command_size
        if command_end > commands_end:
            _reject(f"truncated load command {index}")

        if command == LC_CODE_SIGNATURE:
            if command_size != 16:
                _reject("LC_CODE_SIGNATURE must be exactly 16 bytes")
            dataoff, datasize = _unpack(data, order, "II", command_offset + 8)
            signatures.append(CodeSignature(command_offset, dataoff, datasize))

        if command == expected_segment and _segment_name(data, command_offset + 8) == "__LINKEDIT":
            if command_size < minimum_segment_size:
                _reject("truncated __LINKEDIT segment command")
            if bits == 64:
                _vmaddr, vmsize, fileoff, filesize = _unpack(data, order, "QQQQ", command_offset + 24)
                vmsize_offset = command_offset + 32
                filesize_offset = command_offset + 48
            else:
                _vmaddr, vmsize, fileoff, filesize = _unpack(data, order, "IIII", command_offset + 24)
                vmsize_offset = command_offset + 28
                filesize_offset = command_offset + 36
            linkedits.append(LinkeditSegment(command_offset, "__LINKEDIT", vmsize_offset, filesize_offset, fileoff, filesize, vmsize))
        command_offset = command_end

    if command_offset != commands_end:
        _reject("ncmds does not consume sizeofcmds")
    if len(signatures) != 1:
        _reject(f"expected one LC_CODE_SIGNATURE, found {len(signatures)}")
    if len(linkedits) != 1:
        _reject(f"expected one __LINKEDIT segment, found {len(linkedits)}")
    signature = signatures[0]
    linkedit = linkedits[0]
    if signature.datasize == 0:
        _reject("empty code signature blob")
    signature_end = signature.dataoff + signature.datasize
    linkedit_end = linkedit.fileoff + linkedit.filesize
    if signature_end > len(data) or signature_end < signature.dataoff:
        _reject("code signature blob is out of bounds")
    if linkedit_end > len(data) or linkedit_end < linkedit.fileoff:
        _reject("__LINKEDIT range is out of bounds")
    if signature.dataoff < linkedit.fileoff or signature_end > linkedit_end:
        _reject("code signature blob is outside __LINKEDIT")
    if signature_end != len(data) or linkedit_end != len(data):
        _reject("code signature and __LINKEDIT must end at EOF")
    if signature.dataoff < commands_end:
        _reject("code signature overlaps load commands")

    return MachOImage(bits, endian, header_size, ncmds, sizeofcmds, commands_end, len(data), signature, linkedit)


def normalize_macho(data: bytes) -> bytes:
    parsed = parse_macho(data)
    normalized = bytearray(data[: parsed.signature.dataoff])
    order = "<" if parsed.endian == "little" else ">"
    struct.pack_into(f"{order}II", normalized, parsed.signature.command_offset + 8, 0, 0)
    unsigned_linkedit_size = parsed.signature.dataoff - parsed.linkedit.fileoff
    size_layout = "Q" if parsed.bits == 64 else "I"
    struct.pack_into(f"{order}{size_layout}", normalized, parsed.linkedit.filesize_offset, unsigned_linkedit_size)
    struct.pack_into(f"{order}{size_layout}", normalized, parsed.linkedit.vmsize_offset, unsigned_linkedit_size)
    return bytes(normalized)


def main() -> int:
    if len(sys.argv) != 2:
        raise ValueError("usage: python macho_normalizer.py <path>")
    try:
        normalized = normalize_macho(Path(sys.argv[1]).read_bytes())
    except MachOFormatError as error:
        print(error, file=sys.stderr)
        return 3
    sys.stdout.buffer.write(normalized)
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: exp/exact-replay/js/macho-normalizer.mjs
import { readFileSync } from 'node:fs'

const LC_SEGMENT = 0x1
const LC_SEGMENT_64 = 0x19
const LC_CODE_SIGNATURE = 0x1d

export class MachOFormatError extends Error {
  constructor(message) {
    super(message)
    this.name = 'MachOFormatError'
  }
}

function formatError(message) {
  throw new MachOFormatError(`invalid thin Mach-O: ${message}`)
}

function headerFormat(bytes) {
  if (bytes.length < 4) formatError('truncated magic')
  const magic = bytes.subarray(0, 4).toString('hex')
  if (magic === 'cffaedfe') return { bits: 64, endian: 'little', headerSize: 32 }
  if (magic === 'feedfacf') return { bits: 64, endian: 'big', headerSize: 32 }
  if (magic === 'cefaedfe') return { bits: 32, endian: 'little', headerSize: 28 }
  if (magic === 'feedface') return { bits: 32, endian: 'big', headerSize: 28 }
  formatError(`unsupported magic ${magic}`)
}

function readers(bytes, endian) {
  const little = endian === 'little'
  return {
    uint32(offset) {
      if (offset < 0 || offset + 4 > bytes.length) formatError(`truncated uint32 at ${offset}`)
      return little ? bytes.readUInt32LE(offset) : bytes.readUInt32BE(offset)
    },
    uint64(offset) {
      if (offset < 0 || offset + 8 > bytes.length) formatError(`truncated uint64 at ${offset}`)
      const value = little ? bytes.readBigUInt64LE(offset) : bytes.readBigUInt64BE(offset)
      if (value > BigInt(Number.MAX_SAFE_INTEGER)) formatError(`uint64 at ${offset} exceeds the supported file range`)
      return Number(value)
    },
  }
}

function segmentName(bytes, offset) {
  const end = bytes.indexOf(0, offset)
  const boundedEnd = end === -1 || end > offset + 16 ? offset + 16 : end
  return bytes.subarray(offset, boundedEnd).toString('ascii')
}

export function parseMachO(input) {
  const bytes = Buffer.from(input)
  const format = headerFormat(bytes)
  if (bytes.length < format.headerSize) formatError('truncated header')
  const read = readers(bytes, format.endian)
  const ncmds = read.uint32(16)
  const sizeofcmds = read.uint32(20)
  const commandsEnd = format.headerSize + sizeofcmds
  if (commandsEnd > bytes.length) formatError('load command region exceeds the file')

  const signatures = []
  const linkedits = []
  let commandOffset = format.headerSize
  for (let index = 0; index < ncmds; index += 1) {
    if (commandOffset + 8 > commandsEnd) formatError(`truncated load command ${index}`)
    const cmd = read.uint32(commandOffset)
    const cmdsize = read.uint32(commandOffset + 4)
    if (cmdsize < 8 || cmdsize % 4 !== 0) formatError(`invalid cmdsize for load command ${index}`)
    const commandEnd = commandOffset + cmdsize
    if (commandEnd > commandsEnd) formatError(`truncated load command ${index}`)

    if (cmd === LC_CODE_SIGNATURE) {
      if (cmdsize !== 16) formatError('LC_CODE_SIGNATURE must be exactly 16 bytes')
      signatures.push({
        commandOffset,
        dataoff: read.uint32(commandOffset + 8),
        datasize: read.uint32(commandOffset + 12),
      })
    }

    const expectedSegmentCommand = format.bits === 64 ? LC_SEGMENT_64 : LC_SEGMENT
    if (cmd === expectedSegmentCommand && segmentName(bytes, commandOffset + 8) === '__LINKEDIT') {
      const minimumSize = format.bits === 64 ? 72 : 56
      if (cmdsize < minimumSize) formatError('truncated __LINKEDIT segment command')
      linkedits.push(format.bits === 64
        ? {
            commandOffset,
            segname: '__LINKEDIT',
            vmsizeOffset: commandOffset + 32,
            fileoff: read.uint64(commandOffset + 40),
            filesizeOffset: commandOffset + 48,
            vmsize: read.uint64(commandOffset + 32),
            filesize: read.uint64(commandOffset + 48),
          }
        : {
            commandOffset,
            segname: '__LINKEDIT',
            vmsizeOffset: commandOffset + 28,
            fileoff: read.uint32(commandOffset + 32),
            filesizeOffset: commandOffset + 36,
            vmsize: read.uint32(commandOffset + 28),
            filesize: read.uint32(commandOffset + 36),
          })
    }
    commandOffset = commandEnd
  }

  if (commandOffset !== commandsEnd) formatError('ncmds does not consume sizeofcmds')
  if (signatures.length !== 1) formatError(`expected one LC_CODE_SIGNATURE, found ${signatures.length}`)
  if (linkedits.length !== 1) formatError(`expected one __LINKEDIT segment, found ${linkedits.length}`)
  const signature = signatures[0]
  const linkedit = linkedits[0]
  if (signature.datasize === 0) formatError('empty code signature blob')
  const signatureEnd = signature.dataoff + signature.datasize
  const linkeditEnd = linkedit.fileoff + linkedit.filesize
  if (signatureEnd > bytes.length || signatureEnd < signature.dataoff) formatError('code signature blob is out of bounds')
  if (linkeditEnd > bytes.length || linkeditEnd < linkedit.fileoff) formatError('__LINKEDIT range is out of bounds')
  if (signature.dataoff < linkedit.fileoff || signatureEnd > linkeditEnd) formatError('code signature blob is outside __LINKEDIT')
  if (signatureEnd !== bytes.length || linkeditEnd !== bytes.length) formatError('code signature and __LINKEDIT must end at EOF')
  if (signature.dataoff < commandsEnd) formatError('code signature overlaps load commands')

  return {
    ...format,
    ncmds,
    sizeofcmds,
    commandsEnd,
    fileLength: bytes.length,
    signature,
    linkedit,
  }
}

function write32(bytes, offset, value, endian) {
  if (endian === 'little') bytes.writeUInt32LE(value, offset)
  else bytes.writeUInt32BE(value, offset)
}

function writeSize(bytes, offset, value, parsed) {
  if (parsed.bits === 32) return write32(bytes, offset, value, parsed.endian)
  const bigint = BigInt(value)
  if (parsed.endian === 'little') bytes.writeBigUInt64LE(bigint, offset)
  else bytes.writeBigUInt64BE(bigint, offset)
}

export function normalizeMachO(input) {
  const bytes = Buffer.from(input)
  const parsed = parseMachO(bytes)
  const normalized = Buffer.from(bytes.subarray(0, parsed.signature.dataoff))
  write32(normalized, parsed.signature.commandOffset + 8, 0, parsed.endian)
  write32(normalized, parsed.signature.commandOffset + 12, 0, parsed.endian)
  const unsignedLinkeditSize = parsed.signature.dataoff - parsed.linkedit.fileoff
  writeSize(normalized, parsed.linkedit.filesizeOffset, unsignedLinkeditSize, parsed)
  writeSize(normalized, parsed.linkedit.vmsizeOffset, unsignedLinkeditSize, parsed)
  return normalized
}

if (import.meta.main) {
  const filePath = process.argv[2]
  if (!filePath) throw new Error('usage: bun macho-normalizer.mjs <path>')
  try {
    process.stdout.write(normalizeMachO(readFileSync(filePath)))
  } catch (error) {
    if (!(error instanceof MachOFormatError)) throw error
    process.stderr.write(`${error.message}\n`)
    process.exitCode = 3
  }
}